File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2021/Y21D23.kt
package org.gristle.pdxpuzzles.advent.y2021

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.algorithms.Graph
import org.gristle.pdxpuzzles.utilities.algorithms.Graph.steps

class Y21D23(input: String) : Day {

    companion object {

        /**
         * Utility function takes an index and an element and swaps in that element at the index.
         */
        fun <T> List<T>.swapElementAt(index: Int, newElement: T): List<T> =
            mapIndexed { idx, t -> if (index == idx) newElement else t }
    }

    /**
     * Used to identify the Amphipod and store energy information. Amphipod.E is actually "Empty"
     */
    enum class Amphipod(val energy: Double) {
        A(1.0),
        B(10.0),
        C(100.0),
        D(1000.0),
        E(0.0),
    }

    /**
     * Represents the state of the room.
     */
    data class Room(val id: Amphipod, val spots: List<Amphipod>) {
        val hallIndex = id.ordinal + 1
        val isFinished: Boolean by lazy { spots.all { it == id } }
        val isOpen: Boolean by lazy { spots.all { it == Amphipod.E || it == id } && spots.any { it == Amphipod.E } }
        val isMixed: Boolean by lazy { spots.any { it !in listOf(Amphipod.E, id) } }
        val openings: Int by lazy { spots.count { it == Amphipod.E } }

        val topAmphipod: Amphipod = spots.find { it != Amphipod.E } ?: Amphipod.E

        /**
         * Adds an amphipod to the room.
         */
        fun addAmphipod(): Room {
            val lastEmptyIndex = spots.lastIndexOf(Amphipod.E)
            return copy(spots = spots.swapElementAt(lastEmptyIndex, id))
        }

        /**
         * Removes an amphipod from the room
         */
        fun removeAmphipod(): Room {
            val firstOccupiedIndex = spots.indexOfFirst { it != Amphipod.E }
            return copy(spots = spots.swapElementAt(firstOccupiedIndex, Amphipod.E))
        }
    }

    /**
     * Represents the state of all amphipods in the hallway and rooms.
     */
    data class State(
        val hallway: List<Amphipod>,
        val rooms: List<Room>,
    ) {

        val isEnd = hallway.all { it == Amphipod.E } && this.rooms.all(Room::isFinished)

        fun edges(): List<Graph.Edge<State>> {

            // find open rooms and get corresponding values
            this.rooms.filter(Room::isOpen).forEach { room ->
                // look left
                var steps = 1
                for (hallSpot in room.hallIndex downTo 0) {
                    // look in hall first
                    when (hallway[hallSpot]) {
                        room.id -> return listOf(edgeFromHallToRoom(hallSpot, room, steps)) // found a piece to place
                        Amphipod.E -> {} // keep going
                        else -> break // left blocked; go right
                    }
                    // look in room (if available)
                    val roomSpot = hallSpot - 2
                    if (roomSpot >= 0) {
                        val top = this.rooms[roomSpot].topAmphipod
                        if (top == room.id) {
                            return listOf(edgeFromRoomToRoom(roomSpot, room, steps))
                        }
                    }
                    steps = if (hallSpot == 1) steps + 1 else steps + 2
                }
                // look right
                steps = 1
                for (hallSpot in (room.hallIndex + 1)..hallway.lastIndex) {
                    // look in hall first
                    when (hallway[hallSpot]) {
                        room.id -> return listOf(edgeFromHallToRoom(hallSpot, room, steps))
                        Amphipod.E -> {}
                        else -> break // right blocked; go to pop
                    }
                    // look in room (if available)
                    val roomSpot = hallSpot - 1
                    if (roomSpot <= this.rooms.lastIndex) {
                        val top = this.rooms[roomSpot].topAmphipod
                        if (top == room.id) {
                            return listOf(edgeFromRoomToRoom(roomSpot, room, steps))
                        }
                    }
                    steps = if (hallSpot == 5) steps + 1 else steps + 2
                }
            }

            // Move to popping...
            return buildList {
                this@State.rooms.filter(Room::isMixed).forEach { room ->
                    // look left
                    var steps = 1
                    for (hallSpot in room.hallIndex downTo 0) {
                        if (hallway[hallSpot] == Amphipod.E) add(edgeFromRoomToHall(hallSpot, room, steps)) else break
                        steps = if (hallSpot == 1) steps + 1 else steps + 2
                    }
                    // look right
                    steps = 1
                    for (hallSpot in (room.hallIndex + 1)..hallway.lastIndex) {
                        if (hallway[hallSpot] == Amphipod.E) add(edgeFromRoomToHall(hallSpot, room, steps)) else break
                        steps = if (hallSpot == 5) steps + 1 else steps + 2
                    }
                }

            }
        }

        private inline fun List<Room>.newRooms(id: Amphipod, action: Room.() -> Room): List<Room> = List(size) { idx ->
            if (idx == id.ordinal) {
                rooms[idx].action()
            } else {
                rooms[idx]
            }
        }

        private fun edgeFromRoomToHall(hallSpot: Int, room: Room, steps: Int): Graph.Edge<State> {
            val newAmphipod = room.topAmphipod
            val newState = let {
                val newHallway = hallway.swapElementAt(hallSpot, newAmphipod)
                val newRooms = rooms.newRooms(room.id, Room::removeAmphipod)
                State(newHallway, newRooms)
            }
            val weight = (steps + room.openings + 1) * newAmphipod.energy
            return Graph.Edge(newState, weight)
        }

        private fun edgeFromHallToRoom(hallSpot: Int, room: Room, steps: Int): Graph.Edge<State> {
            val newState = let {
                val newHallway = hallway.swapElementAt(hallSpot, Amphipod.E)
                val newRooms = rooms.newRooms(room.id, Room::addAmphipod)
                State(newHallway, newRooms)
            }
            val weight = (steps + room.openings) * room.id.energy
            return Graph.Edge(newState, weight)
        }

        private fun edgeFromRoomToRoom(roomSpot: Int, room: Room, steps: Int): Graph.Edge<State> {
            val other = this.rooms[roomSpot]
            val newState = let {
                val newRooms = List(rooms.size) { index ->
                    when (index) {
                        room.id.ordinal -> room.addAmphipod()
                        roomSpot -> other.removeAmphipod()
                        else -> rooms[index]
                    }
                }
                copy(rooms = newRooms)
            }

            val weight = (steps + other.openings + 1 + room.openings + 1) * room.id.energy
            return Graph.Edge(newState, weight)
        }

        override fun hashCode(): Int {
            var result = hallway.hashCode()
            result = 31 * result + rooms.hashCode()
            return result
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false

            other as State

            if (hallway != other.hallway) return false
            if (rooms != other.rooms) return false

            return true
        }
    }

    private fun totalEnergy(state: State): Int = Graph
        .dijkstra(
            startId = state,
            endCondition = State::isEnd,
            defaultEdges = State::edges
        ).steps()

    /**
     * Parses strings into State object.
     */
    private fun getState(a: List<String>): State {
        return a.let {
            List(it.size * it.first().length) { i ->
                Amphipod.valueOf(it[i % it.size][i / it.size].toString())
            }.chunked(it.size)
        }.mapIndexed { index, s ->
            val name = Amphipod.entries[index]
            Room(name, s)
        }.let { State(List(7) { Amphipod.E }, it) }
    }

    private val strings = input
        .filter { it in "ABCD\n" }
        .split('\n')
        .filter { it.isNotBlank() }

    override fun part1(): Int = totalEnergy(getState(strings))

    override fun part2(): Int {
        val expandedStrings = strings.dropLast(1) + listOf("DCBA", "DBAC") + strings.last()
        return totalEnergy(getState(expandedStrings))
    }
}

fun main() = Day.runDay(Y21D23::class)

//    Class creation: 2ms
//    Part 1: 14148 (254ms)
//    Part 2: 43814 (532ms)
//    Total time: 789ms

File: rust/advent/src/bin/2021/y21d23.rs
use advent::utilities::get_input::get_input;
use itertools::Itertools;
use rustc_hash::{FxHashMap, FxHashSet};
use std::{
    cmp::{Ordering, Reverse},
    collections::BinaryHeap,
};
use utilities::structs::stopwatch::{ReportDuration, Stopwatch};

type Input = Vec<String>;
type Output = usize;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
struct State {
    hallway: [Amphipod; 7],
    rooms: Vec<Room>,
}

impl State {
    fn is_finished(&self) -> bool {
        self.hallway.iter().all(|&amphipod| amphipod == Amphipod::E)
            && self.rooms.iter().all(|room| room.is_finished())
    }

    fn get_edges(&self) -> Vec<(usize, State)> {
        // find open rooms and get corresponding values
        for room in self.rooms.iter().filter(|room| room.is_open()) {
            // look left
            let mut steps = 1;
            for hall_spot in (0..=room.hall_index()).rev() {
                // look in hall first
                match self.hallway[hall_spot] {
                    amp if amp == room.id => {
                        return vec![self.edge_from_hall_to_room(hall_spot, room, steps)];
                    }
                    Amphipod::E => {} // keep going
                    _ => {
                        break;
                    } // left blocked; go right
                }
                // look in room (if available)
                if let Some(room_spot) = hall_spot.checked_sub(2) {
                    let top = self.rooms[room_spot].top_amphipod();
                    if top == room.id {
                        return vec![self.edge_from_room_to_room(room_spot, room, steps)];
                    }
                }
                steps = if hall_spot == 1 { steps + 1 } else { steps + 2 };
            }
            // look right
            steps = 1;
            for hall_spot in room.hall_index() + 1..7 {
                // look in hall first
                match self.hallway[hall_spot] {
                    amp if amp == room.id => {
                        return vec![self.edge_from_hall_to_room(hall_spot, room, steps)];
                    },
                    Amphipod::E => {}
                    _ => { break; } // right blocked; go to pop
                }
                // look in room
                let room_spot = hall_spot - 1;
                if room_spot < self.rooms.len() {
                    let top = self.rooms[room_spot].top_amphipod();
                    if top == room.id {
                        return vec![self.edge_from_room_to_room(room_spot, room, steps)];
                    }
                }
                steps = if hall_spot == 5 { steps + 1 } else { steps + 2 };
            }
        }

        // move to popping...
        let mut edges = Vec::new();
        for room in self.rooms.iter().filter(|room| room.is_mixed()) {
            // look left
            let mut steps = 1;
            for hall_spot in (0..=room.hall_index()).rev() {
                if self.hallway[hall_spot] == Amphipod::E {
                    edges.push(self.edge_from_room_to_hall(hall_spot, room, steps));
                } else {
                    break;
                }
                steps = if hall_spot == 1 { steps + 1 } else { steps + 2 };
            }
            // look right
            steps = 1;
            for hall_spot in room.hall_index() + 1..7 {
                if self.hallway[hall_spot] == Amphipod::E {
                    edges.push(self.edge_from_room_to_hall(hall_spot, room, steps));
                } else {
                    break;
                }
                steps += if hall_spot == 5 { 1 } else { 2 };
            }
        }
        edges
    }

    fn edge_from_hall_to_room(
        &self,
        hall_spot: usize,
        room: &Room,
        steps: usize,
    ) -> (usize, State) {
        let mut new_hallway = self.hallway;
        new_hallway[hall_spot] = Amphipod::E;
        let new_rooms = Self::new_rooms(&self, room.id, Room::add_amphipod);
        let new_state = State { hallway: new_hallway, rooms: new_rooms };
        let weight = (steps + room.openings()) * room.id.energy();
        (weight, new_state)
    }

    fn edge_from_room_to_room(
        &self,
        room_spot: usize,
        room: &Room,
        steps: usize,
    ) -> (usize, State) {
        let other = &self.rooms[room_spot];
        let new_rooms = (0..self.rooms.len())
            .map(|idx| {
                match idx {
                    i if i == room.id.ordinal() => room.add_amphipod(),
                    i if i == room_spot => other.remove_amphipod(),
                    _ => self.rooms[idx].clone(),
                }
            })
            .collect_vec();
        let new_state = State { rooms: new_rooms, ..*self };
        let weight = (steps + other.openings() + 1 + room.openings() + 1) * room.id.energy();
        (weight, new_state)
    }

    fn edge_from_room_to_hall(&self, hall_spot: usize, room: &Room, steps: usize) -> (usize, State) {
        let new_amphipod = room.top_amphipod();
        let new_state = {
            let new_hallway = self.new_hallway(hall_spot, new_amphipod);
            let new_rooms = self.new_rooms(room.id, Room::remove_amphipod);
            State { hallway: new_hallway, rooms: new_rooms }
        };
        let weight = (steps + room.openings() + 1) * new_amphipod.energy();
        (weight, new_state)
        
    }

    fn new_hallway(&self, index: usize, amphipod: Amphipod) -> [Amphipod; 7] {
        let mut new_hallway = self.hallway;
        new_hallway[index] = amphipod;
        new_hallway
    }

    fn new_rooms<F>(&self, id: Amphipod, action: F) -> Vec<Room>
    where
        F: Fn(&Room) -> Room,
    {
        (0..self.rooms.len())
            .map(|idx| {
                if idx == id.ordinal() {
                    action(&self.rooms[idx])
                } else {
                    self.rooms[idx].clone()
                }
            })
            .collect()
    }
}

impl Ord for State {
    fn cmp(&self, _: &Self) -> Ordering {
        Ordering::Equal
    }
}

impl PartialOrd for State {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
enum Amphipod {
    A,
    B,
    C,
    D,
    E,
}

impl Amphipod {
    const ENTRIES: [Self; 5] = [Self::A, Self::B, Self::C, Self::D, Self::E];
    fn energy(&self) -> usize {
        match self {
            Amphipod::A => 1,
            Amphipod::B => 10,
            Amphipod::C => 100,
            Amphipod::D => 1_000,
            Amphipod::E => 0,
        }
    }

    fn ordinal(&self) -> usize {
        match self {
            Amphipod::A => 0,
            Amphipod::B => 1,
            Amphipod::C => 2,
            Amphipod::D => 3,
            Amphipod::E => 4,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
struct Room {
    id: Amphipod,
    spots: Vec<Amphipod>,
}

impl Room {
    fn is_finished(&self) -> bool {
        self.spots.iter().all(|&it| it == self.id)
    }

    fn is_open(&self) -> bool {
        self.spots
            .iter()
            .all(|&it| it == Amphipod::E || it == self.id)
            && self.spots.iter().any(|&it| it == Amphipod::E)
    }

    fn hall_index(&self) -> usize {
        self.id.ordinal() + 1
    }

    fn top_amphipod(&self) -> Amphipod {
        *self
            .spots
            .iter()
            .find(|&&spot| spot != Amphipod::E)
            .unwrap_or(&Amphipod::E)
    }

    fn is_mixed(&self) -> bool {
        self.spots.iter().any(|&it| ![Amphipod::E, self.id].contains(&it))
    }

    fn add_amphipod(&self) -> Room {
        let last_empty_index = self.spots.iter().rposition(|&it| it == Amphipod::E).unwrap();
        let mut spots = self.spots.clone();
        spots[last_empty_index] = self.id;
        Room { spots, ..*self }
    }

    fn remove_amphipod(&self) -> Room {
        let first_occupied_index = self.spots.iter()
            .position(|&it| it != Amphipod::E)
            .unwrap();
        let mut spots = self.spots.clone();
        spots[first_occupied_index] = Amphipod::E;
        Room { spots, ..*self }
    }

    fn openings(&self) -> usize {
        self.spots.iter().filter(|&&it| it == Amphipod::E).count()
    }
}
fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(21, 23).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(input.clone()), stopwatch.lap().report());
    println!("2. {} ({})", part2(input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn parse_input(input: &str) -> Input {
    input
        .chars()
        .filter(|c| c.is_ascii_alphabetic())
        .chunks(4)
        .into_iter()
        .map(|chunk| chunk.collect())
        .collect()
}

fn total_energy(strings: Vec<String>) -> Output {
    let start = (0, get_state(strings));

    let mut q: BinaryHeap<Reverse<(usize, State)>> = BinaryHeap::new();
    q.push(Reverse(start));

    let mut weights: FxHashMap<State, usize> = FxHashMap::default();
    let mut visited: FxHashSet<State> = FxHashSet::default();

    while let Some(Reverse((weight, state))) = q.pop() {
        if visited.contains(&state) {
            continue;
        }
        if state.is_finished() {
            return weight;
        }

        for (edge_weight, edge_state) in state.get_edges() {
            let potential_weight = weight + edge_weight;
            if let Some(current_weight) = weights.get_mut(&edge_state) {
                if potential_weight < *current_weight {
                    *current_weight = potential_weight;
                    q.push(Reverse((potential_weight, edge_state)));
                }
            } else {
                weights.insert(edge_state.clone(), potential_weight);
                q.push(Reverse((potential_weight, edge_state)));
            }
        }
        visited.insert(state);
    }
    unreachable!("Queue ran out of edges before solution was found!")
}

fn get_state(strings: Vec<String>) -> State {
    let len = strings.len();
    let collated = (0..len * 4)
        .map(|i| match strings[i % len].as_bytes()[i / len] {
            b'A' => Amphipod::A,
            b'B' => Amphipod::B,
            b'C' => Amphipod::C,
            b'D' => Amphipod::D,
            _ => unreachable!(),
        })
        .chunks(len);
    let rooms = collated
        .into_iter()
        .enumerate()
        .map(|(idx, chunk)| {
            let id = Amphipod::ENTRIES[idx];
            Room {
                id,
                spots: chunk.collect(),
            }
        })
        .collect();
    State {
        hallway: [Amphipod::E; 7],
        rooms,
    }
}

fn part1(strings: Input) -> Output {
    total_energy(strings)
}

fn part2(strings: Input) -> Output {
    let (first, last) = strings.into_iter().collect_tuple().unwrap();
    let expanded_strings = vec![first, "DCBA".to_string(), "DBAC".to_string(), last];
    total_energy(expanded_strings)
}

#[test]
fn default() {
    let input = get_input(21, 23).unwrap();
    let input = parse_input(&input);
    assert_eq!(14148, part1(input.clone()));
    assert_eq!(43814, part2(input));
}

// Input parsed (19μs)
// 1. 14148 (44ms)
// 2. 43814 (153ms)
// Total: 198ms